File: apps/studio/src/lib/memory-policies.ts
import { cpJson, type UxWireupClientOptions } from "@/lib/ux-wireup";

export type MemoryPolicyScope =
  | "turn"
  | "conversation"
  | "session"
  | "user"
  | "account"
  | "organization"
  | "task"
  | "agent"
  | "workspace";

export type MemoryPolicyApprovalStatus =
  | "draft"
  | "review_required"
  | "approved"
  | "blocked";

export interface MemoryPolicy {
  id: string;
  workspace_id: string;
  agent_id: string;
  scope: MemoryPolicyScope;
  allowed_memory_types: string[];
  retention: string;
  consent_requirement: string;
  pii_policy: string;
  delete_behavior: string;
  privacy_implications: string[];
  source_trace_required: boolean;
  approval_status: MemoryPolicyApprovalStatus;
  content_hash: string;
  approval_invalidated_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface MemoryPoliciesResponse {
  items: MemoryPolicy[];
}

export type MemoryPolicyInput = Pick<
  MemoryPolicy,
  | "scope"
  | "allowed_memory_types"
  | "retention"
  | "consent_requirement"
  | "pii_policy"
  | "delete_behavior"
  | "privacy_implications"
  | "source_trace_required"
>;

type MemoryPolicyClientOptions = UxWireupClientOptions & {
  allowFixture?: boolean;
};

const LOCAL_NOW = new Date(0).toISOString();

function localPolicy(
  agentId: string,
  input: {
    id: string;
    scope: MemoryPolicyScope;
    allowed_memory_types: string[];
    retention: string;
    consent_requirement: string;
    pii_policy: string;
    delete_behavior: string;
    privacy_implications: string[];
    approval_status: MemoryPolicyApprovalStatus;
  },
): MemoryPolicy {
  return {
    workspace_id: "local",
    agent_id: agentId,
    source_trace_required: true,
    content_hash: `local_${input.scope}_policy`,
    approval_invalidated_at: null,
    created_at: LOCAL_NOW,
    updated_at: LOCAL_NOW,
    ...input,
  };
}

export function localMemoryPolicies(agentId: string): MemoryPolicy[] {
  return [
    localPolicy(agentId, {
      id: "mp_local_session",
      scope: "session",
      allowed_memory_types: ["scratchpad", "conversation_state"],
      retention: "Expires with the active conversation session.",
      consent_requirement: "Covered by channel session disclosure.",
      pii_policy: "PII is kept transient and never promoted to durable memory.",
      delete_behavior: "Session expiry deletes the memory automatically.",
      privacy_implications: [
        "Used only to keep the current task coherent.",
        "Never affects another conversation.",
      ],
      approval_status: "draft",
    }),
    localPolicy(agentId, {
      id: "mp_local_user",
      scope: "user",
      allowed_memory_types: ["preference", "support_context"],
      retention: "Keep confirmed user preferences for 365 days.",
      consent_requirement: "Explicit consent required before durable write.",
      pii_policy:
        "Do not store secrets, payment data, or unconfirmed identifiers.",
      delete_behavior: "Delete on user request with audit trail.",
      privacy_implications: [
        "Durable preference affects future conversations for this user.",
        "Every write must link to the source turn and writer version.",
      ],
      approval_status: "review_required",
    }),
    localPolicy(agentId, {
      id: "mp_local_account",
      scope: "account",
      allowed_memory_types: ["plan_tier", "account_entitlement"],
      retention: "Keep account facts for 90 days unless account data changes.",
      consent_requirement:
        "Admin-approved account memory only; users are notified through account terms.",
      pii_policy: "Account memory cannot include user-level secrets or payment data.",
      delete_behavior: "Delete or re-sync when account ownership or plan changes.",
      privacy_implications: [
        "Account memory can affect all users under the same customer account.",
        "Writes must cite the lookup trace and source system version.",
      ],
      approval_status: "review_required",
    }),
    localPolicy(agentId, {
      id: "mp_local_organization",
      scope: "organization",
      allowed_memory_types: ["routing_policy", "tenant_preference"],
      retention:
        "Retain organization policy memory until superseded by governance.",
      consent_requirement:
        "Workspace admin approval and governance notice required before activation.",
      pii_policy: "Organization memory cannot store individual PII.",
      delete_behavior:
        "Delete through governance request or replace with a superseding policy.",
      privacy_implications: [
        "Organization memory can affect every agent user in the tenant.",
        "Policy changes must appear in deployment preflight.",
      ],
      approval_status: "review_required",
    }),
    localPolicy(agentId, {
      id: "mp_local_task",
      scope: "task",
      allowed_memory_types: ["task_context", "retrieval_context"],
      retention: "Expires when the task or replay branch closes.",
      consent_requirement: "Covered by the active task and trace consent.",
      pii_policy: "Task memory may not be promoted to durable scope automatically.",
      delete_behavior: "Delete when the task is closed or replay context is cleared.",
      privacy_implications: [
        "Task memory only affects the active task, replay, or simulator branch.",
      ],
      approval_status: "draft",
    }),
    localPolicy(agentId, {
      id: "mp_local_agent",
      scope: "agent",
      allowed_memory_types: ["behavior_guardrail", "accepted_catch"],
      retention: "Version-bound retention; supersede with accepted behavior changes.",
      consent_requirement:
        "Builder approval required before agent-scope memory changes behavior.",
      pii_policy: "Agent memory cannot contain user PII or customer secrets.",
      delete_behavior: "Supersede or delete through a governed behavior change.",
      privacy_implications: [
        "Agent memory can affect every conversation handled by this agent.",
        "Changes must link to a catch, incident, eval, or accepted change package.",
      ],
      approval_status: "review_required",
    }),
    localPolicy(agentId, {
      id: "mp_local_workspace",
      scope: "workspace",
      allowed_memory_types: ["team_preference", "operational_exception"],
      retention: "Requires compliance review before any workspace-wide write.",
      consent_requirement: "Admin approval and tenant notice required.",
      pii_policy: "Workspace memory cannot store user-level PII.",
      delete_behavior: "Delete through governance request with audit trail.",
      privacy_implications: [
        "Workspace memory can affect every user in the tenant.",
        "Policy changes must appear in deployment preflight before activation.",
      ],
      approval_status: "blocked",
    }),
  ];
}

export async function listMemoryPolicies(
  agentId: string,
  opts: MemoryPolicyClientOptions = {},
): Promise<MemoryPoliciesResponse> {
  return cpJson<MemoryPoliciesResponse>(
    `/agents/${encodeURIComponent(agentId)}/memory-policies`,
    {
      ...opts,
      allowFallback: opts.allowFixture === true,
      fallback: { items: localMemoryPolicies(agentId) },
    },
  );
}

export async function upsertMemoryPolicy(
  agentId: string,
  input: MemoryPolicyInput,
  opts: MemoryPolicyClientOptions = {},
): Promise<MemoryPolicy> {
  return cpJson<MemoryPolicy>(
    `/agents/${encodeURIComponent(agentId)}/memory-policies/${encodeURIComponent(
      input.scope,
    )}`,
    {
      ...opts,
      method: "PUT",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback: {
        ...localMemoryPolicies(agentId).find(
          (policy) => policy.scope === input.scope,
        )!,
        ...input,
        approval_status:
          input.scope === "user" ||
          input.scope === "account" ||
          input.scope === "organization" ||
          input.scope === "agent" ||
          input.scope === "workspace"
            ? "review_required"
            : "draft",
        updated_at: new Date().toISOString(),
      },
    },
  );
}

export async function approveMemoryPolicy(
  agentId: string,
  scope: MemoryPolicyScope,
  opts: MemoryPolicyClientOptions = {},
): Promise<MemoryPolicy> {
  const fallback =
    localMemoryPolicies(agentId).find((policy) => policy.scope === scope) ??
    localMemoryPolicies(agentId)[0]!;
  return cpJson<MemoryPolicy>(
    `/agents/${encodeURIComponent(agentId)}/memory-policies/${encodeURIComponent(
      scope,
    )}/approve`,
    {
      ...opts,
      method: "POST",
      allowFallback: opts.allowFixture === true,
      fallback: {
        ...fallback,
        approval_status: "approved",
        approval_invalidated_at: null,
        updated_at: new Date().toISOString(),
      },
    },
  );
}

File: packages/control-plane/loop_control_plane/memory_policies.py
from __future__ import annotations

import asyncio
import json
from datetime import UTC, datetime
from hashlib import sha256
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field

from loop_control_plane._app_agents import AgentRecord
from loop_control_plane.workspaces import WorkspaceError

MemoryPolicyScope = Literal[
    "turn",
    "conversation",
    "session",
    "user",
    "account",
    "organization",
    "task",
    "agent",
    "workspace",
]
MemoryPolicyApprovalStatus = Literal["draft", "review_required", "approved", "blocked"]

DURABLE_SCOPES: set[MemoryPolicyScope] = {
    "user",
    "account",
    "organization",
    "agent",
    "workspace",
}


class MemoryPolicyUpsert(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scope: MemoryPolicyScope
    allowed_memory_types: list[str] = Field(default_factory=list, max_length=25)
    retention: str = Field(min_length=1, max_length=500)
    consent_requirement: str = Field(min_length=1, max_length=500)
    pii_policy: str = Field(min_length=1, max_length=500)
    delete_behavior: str = Field(min_length=1, max_length=500)
    privacy_implications: list[str] = Field(default_factory=list, max_length=10)
    source_trace_required: bool = True


class MemoryPolicyRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    workspace_id: UUID
    agent_id: UUID
    scope: MemoryPolicyScope
    allowed_memory_types: list[str]
    retention: str
    consent_requirement: str
    pii_policy: str
    delete_behavior: str
    privacy_implications: list[str]
    source_trace_required: bool
    approval_status: MemoryPolicyApprovalStatus
    content_hash: str
    approval_invalidated_at: datetime | None
    created_at: datetime
    updated_at: datetime


def _content_hash(body: MemoryPolicyUpsert) -> str:
    encoded = json.dumps(
        body.model_dump(mode="json"),
        sort_keys=True,
        separators=(",", ":"),
    )
    return sha256(encoded.encode("utf-8")).hexdigest()


def _approval_status_for(body: MemoryPolicyUpsert) -> MemoryPolicyApprovalStatus:
    if not body.allowed_memory_types:
        return "blocked"
    if not body.source_trace_required:
        return "blocked"
    lowered = " ".join(
        [
            body.consent_requirement,
            body.pii_policy,
            body.delete_behavior,
            *body.privacy_implications,
        ]
    ).lower()
    if body.scope in DURABLE_SCOPES and (
        not body.privacy_implications
        or "delete" not in body.delete_behavior.lower()
        or "consent" not in body.consent_requirement.lower()
    ):
        return "review_required"
    if any(token in lowered for token in ("pii", "personal", "identifier", "email")):
        return "review_required"
    return "draft"


def _approval_blocker(record: MemoryPolicyRecord) -> str | None:
    if not record.allowed_memory_types:
        return "allowed memory types are required"
    if not record.source_trace_required:
        return "durable memory writes must remain trace-backed"
    if record.scope in DURABLE_SCOPES:
        if not record.privacy_implications:
            return "privacy implications must be shown before durable activation"
        if "delete" not in record.delete_behavior.lower():
            return "durable memory requires explicit delete behavior"
        if "consent" not in record.consent_requirement.lower():
            return "durable memory requires explicit consent requirement"
    return None


def memory_policy_payload(record: MemoryPolicyRecord) -> dict[str, Any]:
    return record.model_dump(mode="json")


class MemoryPolicyRegistry:
    def __init__(self) -> None:
        self._items: dict[UUID, dict[MemoryPolicyScope, MemoryPolicyRecord]] = {}
        self._lock = asyncio.Lock()

    async def list_for_agent(self, *, agent: AgentRecord) -> list[MemoryPolicyRecord]:
        async with self._lock:
            return list(self._items.get(agent.id, {}).values())

    async def upsert(
        self,
        *,
        agent: AgentRecord,
        body: MemoryPolicyUpsert,
    ) -> MemoryPolicyRecord:
        async with self._lock:
            now = datetime.now(UTC)
            by_scope = self._items.setdefault(agent.id, {})
            existing = by_scope.get(body.scope)
            content_hash = _content_hash(body)
            approval_invalidated_at = (
                existing.approval_invalidated_at if existing is not None else None
            )
            approval_status = _approval_status_for(body)
            if (
                existing is not None
                and existing.approval_status == "approved"
                and existing.content_hash != content_hash
            ):
                approval_invalidated_at = now
                approval_status = "review_required"

            record = MemoryPolicyRecord(
                id=existing.id if existing is not None else f"mp_{uuid4().hex[:12]}",
                workspace_id=agent.workspace_id,
                agent_id=agent.id,
                scope=body.scope,
                allowed_memory_types=body.allowed_memory_types,
                retention=body.retention,
                consent_requirement=body.consent_requirement,
                pii_policy=body.pii_policy,
                delete_behavior=body.delete_behavior,
                privacy_implications=body.privacy_implications,
                source_trace_required=body.source_trace_required,
                approval_status=approval_status,
                content_hash=content_hash,
                approval_invalidated_at=approval_invalidated_at,
                created_at=existing.created_at if existing is not None else now,
                updated_at=now,
            )
            by_scope[body.scope] = record
            return record

    async def approve(
        self,
        *,
        agent: AgentRecord,
        scope: MemoryPolicyScope,
    ) -> MemoryPolicyRecord:
        async with self._lock:
            record = self._items.get(agent.id, {}).get(scope)
            if record is None:
                raise WorkspaceError(f"unknown memory policy scope: {scope}")
            blocker = _approval_blocker(record)
            if blocker is not None:
                raise WorkspaceError(blocker)
            updated = record.model_copy(
                update={
                    "approval_status": "approved",
                    "approval_invalidated_at": None,
                    "updated_at": datetime.now(UTC),
                }
            )
            self._items[agent.id][scope] = updated
            return updated
